jose/x25519_test.go

```diff
@@ -0,0 +1,360 @@
+			ID:        "test-jti",
+		}, false},
+		{"ok empty", args{
+			SigningKey{
+				Algorithm: XEdDSA,
+				Key: x25519.PrivateKey{
+					0x4e, 0x6c, 0xd9, 0xe8, 0x40, 0x4d, 0xe1, 0xe0,
+					0xc5, 0x22, 0x72, 0x8f, 0x78, 0xde, 0x88, 0x26,
+					0x58, 0x4b, 0x44, 0xef, 0xbc, 0xf6, 0xac, 0x10,
+					0x97, 0x67, 0x60, 0xcc, 0x41, 0xf1, 0x82, 0x0b,
+				},
+			},
+			new(SignerOptions).WithType("JWT"),
+			[]byte(`{}`),
+			x25519.PublicKey{
+				0x51, 0xf3, 0xfe, 0x72, 0xb4, 0x35, 0x66, 0x52,
+				0x16, 0x16, 0x38, 0xfb, 0x9b, 0xf7, 0x17, 0x06,
+				0x87, 0xb6, 0x35, 0x53, 0xc1, 0x63, 0x9c, 0x73,
+				0xa7, 0x79, 0x1e, 0xd2, 0xba, 0xf8, 0xcb, 0x67,
+			},
+		}, Claims{}, false},
+		{"ok random", args{
+			SigningKey{Algorithm: XEdDSA, Key: priv},
+			new(SignerOptions).WithType("JWT"),
+			[]byte(`{"iss":"test-iss","sub":"test-sub","aud":"test-aud","exp":1234,"nbf":1200,"iat":1000,"jti":"test-jti"}`),
+			pub,
+		}, Claims{
+			Issuer:    "test-iss",
+			Subject:   "test-sub",
+			Audience:  []string{"test-aud"},
+			Expiry:    NewNumericDate(time.Unix(1234, 0)),
+			NotBefore: NewNumericDate(time.Unix(1200, 0)),
+			IssuedAt:  NewNumericDate(time.Unix(1000, 0)),
+			ID:        "test-jti",
+		}, false},
+		{"ok ed25519", args{
+			SigningKey{Algorithm: EdDSA, Key: edPriv},
+			new(SignerOptions).WithType("JWT"),
+			[]byte(`{"iss":"test-iss","sub":"test-sub","aud":"test-aud","exp":1234,"nbf":1200,"iat":1000,"jti":"test-jti"}`),
+			edPub,
+		}, Claims{
+			Issuer:    "test-iss",
+			Subject:   "test-sub",
+			Audience:  []string{"test-aud"},
+			Expiry:    NewNumericDate(time.Unix(1234, 0)),
+			NotBefore: NewNumericDate(time.Unix(1200, 0)),
+			IssuedAt:  NewNumericDate(time.Unix(1000, 0)),
+			ID:        "test-jti",
+		}, false},
+		{"fail algorithm", args{
+			SigningKey{Algorithm: EdDSA, Key: priv},
+			new(SignerOptions).WithType("JWT"),
+			[]byte(`{"iss":"test-iss","sub":"test-sub","aud":"test-aud","exp":1234,"nbf":1200,"iat":1000,"jti":"test-jti"}`),
+			pub,
+		}, Claims{}, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			signer, err := NewSigner(tt.args.sig, tt.args.opts)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewSigner() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !tt.wantErr {
+				var got Claims
+
+				sig, err := signer.Sign(tt.args.payload)
+				if err != nil {
+					t.Errorf("Signer.Sign() error = %v", err)
+					return
+				}
+				jwt, err := ParseSigned(sig.FullSerialize())
+				if err != nil {
+					t.Errorf("ParseSigned() error = %v", err)
+					return
+				}
+				if err := Verify(jwt, tt.args.publicKey, &got); err != nil {
+					t.Errorf("Verify() error = %v", err)
+					return
+				}
+				if !reflect.DeepEqual(got, tt.want) {
+					t.Errorf("JSONWebSignature.CompactSerialize() = %v, want %v", got, tt.want)
+				}
+			}
+		})
+	}
+}
+
+func TestX25519Signer_Public(t *testing.T) {
+	pub, priv, err := x25519.GenerateKey(rand.Reader)
```

Comment:
While investigating the deterministic output for `rand.Reader`, I saw that the Go stdlib implementation of `ed25519.GenerateKey` checks if `rand.Reader` is nil, and if so, will fallback to `crypto/rand.Reader`. It might be nice to mirror that approach for `x25519.GenerateKey`, unless you expect problems down the road?

Reply:
I prefer not to do that, as it can hide other errors, it will probably panic if it's nil.